### ISSUE TITLE
Use programmatic content scripts?

### ISSUE THREAD

**Issue by arantius** [greasemonkey/greasemonkey#2663]
https://bugzilla.mozilla.org/show_bug.cgi?id=1332273

Mostly setting this up to make sure the bugzilla bug is tracked.  This could/should be useful for us, especially for `@document-start` support?  Look into this after the above bug is resolved.  If it's useful, gate on stable release of the feature and implement.

**Comment by Sxderp** [comment]
Assuming I understand it correctly, could be useful in that GM doesn't have to figure out what scripts match on navigation, only when installing / modifying and then have Firefox do the rest.

Though, for the topic of `@document-start` does switching to [webRequest.onRequestStarted](https://github.com/greasemonkey/greasemonkey/pull/2635) have any side effects in regard to being more consistent? I don't know how that can be tested / checked.

**Comment by arantius** [comment]
> ... for the topic of @document-start does switching to webRequest.onRequestStarted ...

Even if we can detect the execution time, we have to (async) beam the script content down, which arrives at some non-start time.  In theory if we register the script with Firefox, it will "just work"?

(I don't want to get in the business of pre-caching things in content scopes in the hopes that it will already be there at start time.)

**Comment by kekkc** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1332273 was fixed in latest FF59 some hours ago:
> Keywords: dev-doc-needed
Summary: Pass initial options to content scripts without async messaging (contentScriptOptions equivalent) → Implement a contentScripts.register API

So we need an additional API to use this?

**Comment by Sxderp** [comment]
> So we need an additional API to use this?

No. I don't think so. They where just modifying the description / summary in order to more clearly label what the bug was about. Documentation is needed though. But looking at the [test cases](https://hg.mozilla.org/integration/autoland/diff/54ce507fdd65/toolkit/components/extensions/test/xpcshell/test_ext_contentScripts_register.js) can give an idea on how this is implemented in a WebExtension.

As for Greasemonkey, I think this should allow for the removal of the `execute.js` file. And we just register the scripts when they're first loaded / saved.

**Comment by Sxderp** [comment]
I took a quick look at this and it'll be a bit tricky in regard to matches and includes. Greasemonkey takes the 'OR' approach while Mozilla [content scripts](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json/content_scripts#Matching_URL_patterns) take an 'AND' approach for matches + includes. This is doable but not very cleanly.

Also regex not supported.

**Comment by Sxderp** [comment]
I have a POC up but it doesn't have regex support (sigh). And I think it looks good.

Some issues that it resolves:
#2574 and the problems in my initial pull #2635
#2705
Partially #2715, (monkey menu still needs an update to list proper scripts).

**Comment by kekkc** [comment]
Sxderp, you're awesome ;)

Can't really review code, but I've seen you've linked to a manifest.json page. Regarding the match / regex problem I just found the related site. Sounds to me like programmatically can be used for all tabs. Maybe this could include then an additional Regex check for the URL?
https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Content_scripts

> You can load a content script into a web page in one of two ways:
>1. declaratively: 
 using the content_scripts key in your manifest.json, you can ask the browser to load a content script whenever the browser loads a page whose URL matches a given pattern
>2. programmatically: 
using the tabs.executeScript() API, you can load a content script into a specific tab whenever you want: for example, in response to the user clicking on a browser action.

**Comment by Sxderp** [comment]
> Maybe this could include then an additional Regex check for the URL?

The method described on that page is what GM currently uses, `tabs.executeScript`. And it has a slew of associated with it (see above).

I'm not sure if regex is considered a major dealbreak or not. I don't use it, sticking with globs, but others might use it.

**Comment by NoneGiven** [comment]
@Sxderp You can always just bail out of your script after testing the URL against a regex yourself.

Also, #2715 is really bothersome. Is it blocked on this or can we just quickly apply the fix described by @arantius in that issue?

**Comment by kekkc** [comment]
>I have a POC up but it doesn't have regex support
Do you have a testcase? BTW: there's a bug in FF's test to register scripts ( https://bugzilla.mozilla.org/show_bug.cgi?id=1419576 )

**Comment by Sxderp** [comment]
@kekkc should be one of my branches. Clone, checkout, load temporary add-on.

**Comment by kekkc** [comment]
@Sxderp pretty cool, it worked with FF59 nightly, no errors in console. I tested a userscript that just alerts sth. and used a website that also just alerts sth. 
--> my userscript alert was displayed before anything from the website was loaded and the website alert was displayed afterwards
--> this is the first time in FF and GM history that we're able to execute our scripts before anything else (all my other scripts using mutationobservers or overriding alert/confirm JS methods are correctly working again), historic moment for me ;)

Only one small drawback: 
- seems that unregister is not direcly working in your branch https://github.com/Sxderp/greasemonkey/tree/use-programmatic-content-scripts : 
I uninstalled my alert userscript via monkeymenu and it was no longer visible after some seconds. However, I still got the alert msg whenever I surfed & loaded a new website (still no errors in console, though).
- Secondly, I think your commit can only be used in GM master when FF59 arrives. I guess FF is https://bugzilla.mozilla.org/show_bug.cgi?id=1332273 not porting back to FF57/58 right?

**Comment by trlkly** [comment]
It would be nice to keep regex support for existing \@include usage. It's the main feature that makes \@include useful rather than just having \@match, IMO. 

The most obvious implementation, to me, is to just have such script match all URLs (unless \@match is also present) and then do a quick check to see if it matches the Regex, returning early if it doesn't. 

With the script being available at document-start, I don't think this would be a significant performance degradation.

**Comment by Sxderp** [comment]
> seems that unregister is not direcly working in your branch ..

Ouch, thought I fixed all the issues regarding unregistered. Though, there are some changes I want to make, so this will probably get wrapped up in those changes.

> Secondly, I think your commit can only be used in GM master when FF59 arrives.

Correct.

**Comment by Sxderp** [comment]
> The most obvious implementation, to me, is to just have such script match all URLs (unless @match is also present) and then do a quick check to see if it matches the Regex, returning early if it doesn't.

While this is possible, just have some wrapper code, it'll also defeat part of the benefit that is gained from using the programmatic content script registration. All page matching is handled by Firefox. If there are no other *@match* or *@include* then it's reasonable to do a check for the regex. However, once we start adding additional flags then you'd have to re implement the entire matching logic in the userscript. Otherwise this becomes fuzzy with figuring out what to match against.

I suppose you could do separate registrations with different wrapper code for the regex, but that feels bad to me.

It's not a very nice situation.

**Comment by NoneGiven** [comment]
I'm in favor of accepting no regex support and letting scripts add their own oneliner for a regex test and return if they want to.

**Comment by xor10** [comment]
@Sxderp Cool! The user scripts appear in the debugger's sources panel. 

-------------
Small fix:
```diff
diff --git a/src/user-script-obj.js b/src/user-script-obj.js
index 0aa7375..ef4c3bf 100644
--- a/src/user-script-obj.js
+++ b/src/user-script-obj.js
@@ -285,7 +285,7 @@ window.EditableUserScript = class EditableUserScript
         userScript();
         })();
         } catch (e) { console.error("Script error: ", e); }
-        //# sourceURL=user-script:${escape(this.id)}`;
+        `;
     this._evalContentVersion = EVAL_CONTENT_VERSION;
   }
 

```

**Comment by trlkly** [comment]
Oh! Now I see the problem. I was assuming that the included regex code was supposed to run after the matches and other includes had completed. So it would be `(match || include_glob) && include_regex`. But the way it actually worked in previous versions was `match || include_glob || include_regex`.  

In that case, I think what would be useful would be to see if we can find some data on how many scripts from the userscript sites (e.g. openuserjs.org, greasyfork.org, and possibly an archive of userscripts.org)  actually use the regex feature. Maybe we could change it to work as`(match || include_glob) && include_regex` without hurting much. Or we could say that, if you use one regex, then all other uses of `@include` must also be regex, and `@match` is not allowed. (I prefer this option myself, barring a perfect solution. The regex scripts I have all only use regex.)

Excludes, of course, are easy. They actually are `(match || include) && !exclude`. It's trivial to test early on if the exclude value is a regex, and then bail out of the script if the regex matches the URL. Heck, there's another possible option: only `@exclude` is allowed to use regex. 

My ideal is that we break as few scripts as reasonably possible. I know the promise-based system broke GM_ functions, but many scripts don't use those.

**Comment by trlkly** [comment]
@noneGiven: That's fine for new code, but what about preexisting code? I think we'd do better to find out how often regex is used with `@include`. 

It's better to check how much you will break before choosing the most breaking solution.

Question is, does anyone know how to try to find such data, without crawling through the UserScript websites ourselves?

**Comment by NoneGiven** [comment]
@trlkly It's already broken for anyone using GM4 right now.

If it absolutely must be implemented, then it needs to be done so I don't take any perf hits unless I actually use `@include` with a regex.

**Comment by Sxderp** [comment]
> @kekkc, seems that unregister is not direcly working in your branch

Should work now for script enable / disable, script uninstall, and global enable / disable.

> @xor10, Cool! The user scripts appear in the debugger's sources panel.
Small fix:

Included

---
@trlkly, even allowing regex for `@exclude` will remove some of the benefit of using the programmatic method. But, adding a wrapper for `@exclude` into the script is indeed much easier than for `@include`.

Honestly, I'd like what @NoneGiven suggested. Let authors implement the Regex. There's not going to be any performance improvements if those details are pushed onto GM. However, there's probably going to be backlash, a la #2724, and the many times it has come up.

It's also important to note, FF59 is still a few months off.
https://wiki.mozilla.org/RapidRelease/Calendar

**Comment by Sxderp** [comment]
So, I looked at all my scripts (most of them I've made), and only one has any regex used. On an excludes. `/.*\.(jpe?g|png|gif|ico)(\?|$)/`
Honestly, not that complex and could be made into a multi line exclude.

**Comment by kekkc** [comment]
>@Sxderp Should work now for script enable / disable, script uninstall, and global enable / disable.
Works, tested with several registers & unregisters via MonkeyMenu. No error in Console & all scripts were correctly working. Fantastic ;) 

>It's also important to note, FF59 is still a few months off.
Wow, 2018-03-13. Maybe @arantius can implement this commit in Beta (probably after GM4.1) and set the min version to 59. I asked at bugzilla if the plan to backport this to FF58.

**Comment by Sxderp** [comment]
Looking back at some Bugzilla entries; https://bugzilla.mozilla.org/show_bug.cgi?id=1391669 was marked as duplicate of https://bugzilla.mozilla.org/show_bug.cgi?id=1332273 but there isn't any privilege restriction options in that API.

Might be worth a new bug with the explicit goal of adding restrictions to the `contentScripts` api.

**Comment by kekkc** [comment]
>Might be worth a new bug with the explicit goal of adding restrictions to the contentScripts api.
Sounds like Cu.Sandbox https://bugzilla.mozilla.org/show_bug.cgi?id=1353468 . Wondering if when we can debug our scripts again (and click on a link in a debug message to highlight the affected source code line).

**Comment by Sxderp** [comment]
> Sounds like Cu.Sandbox ...

Similar, but I don't think that is going to happen. To me, directly creating a Cu.Sandbox seems fairly broad and not in line with the goals Mozilla has for WebExtensions. Adding new methods to an existing API (although new) that modify the functionality of the API seems much more likely of being implemented. While both provide _roughly_ the same outcome I feel this way has better semantics which seem pretty important in the WebExtension ecosystem.

> Wondering if when we can debug our scripts again

Posted somewhere in this thread, using the new `contentScript` API makes the scripts debuggable. Though, I haven't done any specific testing.

**Comment by kekkc** [comment]
>Posted somewhere in this thread, using the new contentScript API makes the scripts debuggable.

Yes, it works. But there seem to be some quirks. If the userscript contains an error the following is displayed in the console and one can click on the storage number part (great):
`SyntaxError: unterminated string literal 34980f5e-62c7-4c3d-905b-4d23c66e40b2:6:6`

The error is then highlighted in the source editor (again, fantastic). However, my test script just contained the alert line plus the comment headers. Everything else displayed in the screenshot comes from GM and is not part of the userscript. I guess it's the old error message handling that shouldn't be displayed here:
![scriptdebug](https://user-images.githubusercontent.com/6283064/34342872-95e2aa46-e9be-11e7-84ef-a4baedde8a81.png)

BTW: happy Xmas to everyone ;)

**Comment by Sxderp** [comment]
> I guess it's the old error message handling that shouldn't be displayed here:

No. It's actually the entire user script. All the 'extra bits' you see are added in order to provide API functionality as well as `@require` and `@resource` data. When you edit the script all of these bits are hidden from you, but they're indeed required to run the userscript.

Now, if you want it removed so that the error page only shows _your_ code, well I don't think that's possible.

**Comment by kekkc** [comment]
OK, I was just surprised and thought maybe the formatting could be optimized (2 blanks before "// ==UserScript==" or even a comment like "UserScript start" & "UserScript end"), but that's minor. Meat we can use this at all in Beta with programmatic content scripts.

**Comment by trlkly** [comment]
I have no inherent problem with losing regex support. I've already fixed up my one script that used regex. I just load it on all pages of the site but have it fail if certain elements aren't present on the page. (Scripts being wrapped in a function makes that easy.)

My issue with making the script authors do it is simply that I don't want a bunch of scripts to break unnecessarily. That's why I was hoping someone knew of a way to check how much breakage we'd have. If nothing else, contact the major userscript hosts and see if they know. We just had a hugely breaking change, and I think it would be good to minimize others. 

If the breakage is small, then remove it. I just think the final decision needs to be made as soon as possible, so there's enough time to warn script authors. 

It would also be good to give a generic guide on the Wiki for users with scripts they rely upon that are no longer being maintained. For the simple cases where all the includes and excludes were regexes, it's a simple one-liner, plus simple changes to the meta block. 

Heck, you can even exploit the data URI scheme to run the test before any further `@require` scripts are run, e.g. put the following before any other `@require`:

    // @require data:text/javascript,if(!/include1|2|3/.test(document.URL)||/exclude1|2|3/.test(document.URL))return

**Comment by trlkly** [comment]
Quick question: might this fix an issue where scripts do not run on the currently open tab after a restart when updating? Or should I file that as a separate issue?

**Comment by trlkly** [comment]
I just now read the content scripts link above, and I see a potential problem. 

> content scripts cannot see JavaScript variables defined by page scripts

A lot of the more advanced scripts I've encountered will deliberately use some of the page's own functions. I've even written one that does this. Can this issue be mitigated--ideally without APIs like unsafeWindow?

**Comment by Sxderp** [comment]
> might this fix an issue where scripts do not run on the currently open tab after a restart when updating?

This is probably related to the below:
https://github.com/greasemonkey/greasemonkey/issues/2729#issuecomment-347036483

> I just now read the content scripts link above, and I see a potential problem.

Scripts in 4.x are already considered 'content scripts' and therefore get Xrays by default. You can still access the page variables using `window.wrappedJSObject`.

**Comment by arantius** [comment]
We should be moving away from user-scripts-as-content-scripts; see  https://bugzilla.mozilla.org/show_bug.cgi?id=1353468

**Comment by Sxderp** [comment]
> We should be moving away from user-scripts-as-content-scripts

I believe it's still the same API calls, just with an additional flag for reducing permissions / different sandbox level. I think my initial POC is still along the same lines of what will be needed going forward.

**Comment by Sxderp** [comment]
@arantius, What's the ruling for keeping, or not keeping, Regex support in `@include` and `@exclude`? FF 59 is approaching and I assume we'll be switching over to the `contentScript` API. [1][2] As mentioned earlier, the APIs do not support Regex matching of URLs, only match patterns and globs. To keep Regex moving forward, some options come to mind (alternatives would also be good). They all suck. 

1. Use the same system we currently have but tailored just for Regex.
   - Has all the same problems we currently have with script execution.
   - To minimize the chance of the script being executed twice we have to do additional checks to ensure globing and / or match haven't already run the script. [3]
2. Do all checking in the header of the user script and return if it doesn't match.
   - Have to 'run' all user scripts on every page.
   - Lose the efficiency of letting Firefox do the matching (doubly so since I believe they're moving `MatchPattern` to C code).
3. Only support Regex for `@exclude` by performing a check in the script header.
   - Slightly more efficient since we don't have to 'run' every script on every web page. Only pages that have already passed a match.
   - Could be confusing and, to me at least, seems a bit weird from a user perspective to support `@exclude` Regex but not `@include`.

[1] Which we should as it solves a lot of problems like `document_start`, scripts being re-executed on forward / back navigation, and frames.
[2] And eventually `userScript` API when it's released down the line.
[3] Re-executing the script has to be reexamined in the context of using the `contentScript` API. Since the API uses `[@match x OR @match y] AND [@include x OR @include y]` scripts are registered twice, once for `@matches` and the other for `@includes`. This may cause double script execution. [4]
[4] Assuming [3] does cause double execution one option would be to convert `@match` into globs for use in the API. If this can be done safely and without causing additional (or reduced) matches, I think it's viable.

**Comment by arantius** [comment]
> FF 59 is approaching and I assume we'll be switching over to the contentScript API.

I have no plan to, given https://bugzilla.mozilla.org/show_bug.cgi?id=1353468 / https://wiki.mozilla.org/WebExtensions/UserScripts

**Comment by Sxderp** [comment]
There's no reason not to switch. We don't know when the userScript API is going to land. And from the looks of it, the userScript API is going to use almost the same interface as the current contentScript API. So we'd be fixing a lot of issues as well as setting up for easy migration. 

Regardless of those benefits, my issues still stand. It looks like the userScript API is going to use the same matching mechanism as the contentScript. Thus regex issues.

**Comment by quilloss** [comment]
`Do all checking in the header of the user script and return if it doesn't match.` only when a regex is present. This way only scripts with regex gets the efficiency loss. I would guess the vast majority of scripts do not use regex matching.

**Comment by welwood08** [comment]
How feasible is it to convert a regex into a match pattern that can be guaranteed to match every input that the regex would have and just deal with any unwanted matches by checking the regex properly when the userscript runs? This way it's not quite as bad as running the userscript on every page (unless the regex converts into `*` or something).

**Comment by Sxderp** [comment]
@welwood08, Not really feasible. Regex can be quite complex, and even have conditionals. I think that would be prone to more errors and bugs.

**Comment by trlkly** [comment]
Yeah, even partial support (example, just limiting to the right host name) would be error prone. There's a reason why regex is directly supported in most languages, rather than making people roll their own. 

Still, that does give me an idea. Since the userScript API hasn't been implemented yet, it might be useful to see if we can get them to also handle regex. RegExp objects are first-class JavaScript entities, after all. At the very least, it could be brought up in the bug in question as a possibility. 

It could be a part of step 3: further API enhancements, so that it woudln't delay implementation of the first two, which are most important.

**Comment by Sxderp** [comment]
I feel it's unlikely to get implemented in any reasonable time frame. I bet the backend matching code is the same for all instances that use matching. Therefore any changes to it would probably require the entire nightly beta release cycle.

**Comment by Sxderp** [comment]
Looks like we have another problem.

> Since the API uses `[@match x OR @match y] AND [@include x OR @include y]` scripts are registered twice ... This may cause double script execution.

I just tested this. Double execution of scripts does occur. Unfortunately my plan to convert `@match` into similar globs doesn't work out. Globs are greedy and therefore you cannot independently match the domain and paths.

For example: `@match http://*.example.net/file`, which should match any subdomain of `example.net` with path `file`. Turning this into a glob would result in `@include http://*.example.net/file`. I tested on my own domain and found that particular include would match on `http://ncgamers.org/files/baz/foo.example.net/file`.

I'm going to inquire on Bugzilla about adding some flag for changing the "AND" to an "OR". I don't think it'll be possible to get Regex support in a reasonable timeframe, but this may be doable.

**Comment by Eselce** [comment]
`@include http://*.example.net/file` seems wrong to me (not just because of `.*`.
Shouldn't it at least be `@include http://[^/]+.example.net/file`?

**Comment by Eselce** [comment]
`@include http://*.example.net/file` seems wrong to me (not only because of `.*`).
Shouldn't it at least be `@include /^http://[^/]+\.example\.net/file$/`?

**Comment by Sxderp** [comment]
@Eselce that's using Regex. The script APIs don't support  regex and therefore I was attempting to do it through globs only. If I have to resort to regex for `@match` then there's little benefit using the built in matching mechanism and the matching might as well be done entirely in the script header.

**Comment by Eselce** [comment]
Ah, okay, I thought, you tried to convert the pattern to RegEx in order to eliminate the `AND`. Misinterpreted the globs part (and the @include pattern as GM-style pattern)...

**Comment by kekkc** [comment]
Hm, the reason why I joined github was https://github.com/greasemonkey/greasemonkey/issues/1849 in 2013. 4 years later now, there's a reliable solution to execute scripts at start (with the involvment of over 30 people on https://bugzilla.mozilla.org/show_bug.cgi?id=1332273 ), it's also the preferred one on https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Content_scripts and we have a working PR from Sxderp. After FF57 the first chance to be not only feature comparable, but better than GM3.

@arantius 
will this land for FF 59 release?

**Comment by Sxderp** [comment]
@kekkc , unfortunately there are some issues to work out in regard to matching URLs to run the script on. So while this change can resolve a lot of outstanding issues with 4.x, it regresses the use of `@include` and `@match`. Still trying to figure these issues out.

**Comment by Sxderp** [comment]
For anyone that wants to test, I've updated my [branch](https://github.com/Sxderp/greasemonkey/tree/use-programmatic-content-scripts).

It's not pretty but it should support regex in `@include` and `@exclude` as well as support having both `@match` and `@include`.

Logic is as follows:
- If there is a regex in `@excludes` goto 'header-match'
- If `<all_urls>` is found in `@matches`  goto 'api-match'
- If there is a regex in `@includes` goto 'header-match'
- If `@matches` > 0 and `@includes` > 0 goto 'header-match'
- Else goto 'api-match'

**header-match**
All `@includes` and `@matches` are converted to regular expressions and appended together using alteration ( `|` ). The same applies for `@excludes`. Then a code snippet is injected into the header of the script which will perform the url matches.

**api-match**
Uses the match patterns and globs available with the API. No code is injected into the script for matching.

**Comment by Sxderp** [comment]
_Sigh_, my implementation doesn't account for https://github.com/greasemonkey/greasemonkey/issues/2795. Why is `@include` so overloaded..

**Comment by Eselce** [comment]
`@include` is a powerful tool. I like that! ;-)

**Comment by welwood08** [comment]
Since when was `<all_urls>` a valid value for `@match`?

**Comment by Sxderp** [comment]
It's a valid value according to [match pattern specifications](https://developer.chrome.com/extensions/match_patterns). Why would it not be valid? It's equivalent to `@include *`.

**Comment by welwood08** [comment]
Yes, apologies, I'm half awake and already removed my comment after I found that!

**Comment by trlkly** [comment]
I believe that should be `*`, which was the Greasemonkey API equivalent to
<all_urls>.

Either way, I'd suggest also handling `@match *://*/*` the same way, as it
is semantically the same. It's what I always used in my scripts.

**Comment by Sxderp** [comment]
> Either way, I'd suggest also handling `@match *://*/*` the same way, as it is semantically the same.

That's not quite the same. `@match *://*/*` will only match `http` and `https`. However, `<all_urls>` should match `file:///*`.

Working on a few adjustments. Like the returning for loop and adding `.tld` support.

**Comment by Sxderp** [comment]
Alright, pushed an update for `.tld`.

Logic is as follows:
- If there is a regex or `.tld` in `@excludes` goto 'header-match'
- If `<all_urls>` is found in `@matches`  goto 'api-match'
- If there is a regex or `.tld` in `@includes` goto 'header-match'
- If `@matches` > 0 and `@includes` > 0 goto 'header-match'
- Else goto 'api-match'

(scroll up for the other bits)

**Comment by welwood08** [comment]
If `@exclude` has a regex but `@include` does not, would it not be more performant to combine api-match and header-match modes? i.e. use the api for includes/matches and then use the header for excludes.

**Comment by Eselce** [comment]
You are quite busy today... ;-)

**Comment by Sxderp** [comment]
@welwood08 , Probably. However, I'll have to think on that and see how it affects the already complex code. If I think of a clean way to integrate it I'll do it, but for now it might be best to stick with an all or nothing approach.

**Comment by Sxderp** [comment]
So, from the kind folks over at OpenUserJS.org and Greasyfork I got my hands on a bunch of script meta data. Here are the results.

```
Total of of scripts: 5591

Have about:blank
#:  0
%:  00.00
Have both match and include
#:  197
%:  03.52
Have Regex (include or exclude)
#:  222
%:  03.97
Have match and Regex (include or exclude)
#:  5
%:  00.08
Have Regex in include:
#:  211
%:  03.77
%%: 95.05
Have Regex in exclude:
#:  32
%:  00.57
%%: 14.41
```

I posted further analysis over in the Bugzilla ticket.
https://bugzilla.mozilla.org/show_bug.cgi?id=1437098#c4

**Comment by sizzlemctwizzle** [comment]
@Sxderp 
Regex patterns aren't used often because they are almost always needlessly complicated for url matching (which is why GM converts globs to regex internally). The feature isn't widely known, and I'm sure most script authors are uncertain about compatibility. Any rare complicated logic could just be implemented by the script or some library.

As far as the `.tld` is concerned, it has been my understanding (I don't much about it after all this time with GM) that it is a workaround for the greediness of wildcards in globs. Like how `https://*.google.*/` would also match when `google` is a subdomain  (e.g. `https://mail.google.example.com/`). Some domains, like `google` are registered with many TLDs, and so `.tld` acts as a shortcut so you don't need an `@include` for each TLD. If there are some additional uses of it, I have no idea. I'm not sure how useful this feature is, how many scripts actually use/need it, or even how many script authors know about it (or even what it's for).

Anyway, that is just my take on things. I'd recommend just going with what is supported by the API for url matching.

**Comment by kekkc** [comment]
>TLD. If there are some additional uses of it, I have no idea. I'm not sure how useful this feature is, how many scripts actually use/need it, or even how many script authors know about it (or even what it's for).

Alot: https://bugzilla.mozilla.org/show_bug.cgi?id=1315558#c14 ;)

**Comment by Sxderp** [comment]
> Like how https://*.google.*/ would also match when google is a subdomain (e.g. https://mail.google.example.com/). Edit: _Does this problem apply to `@match` patterns as well?_

No. Match patterns do not have that problem. Namely because it's an invalid match pattern. Wildcard can only be used for subdomains for matches.

**Comment by kekkc** [comment]
>Have Regex in include %:  03.77
>Any rare complicated logic could just be implemented by the script or some library.

So how do we come to a conclusion here? 
- RegEx just for include
- RegEx only via script (or use userscripts directly and implement possible outcome of https://bugzilla.mozilla.org/show_bug.cgi?id=1437098#c4 later)

**Comment by Eselce** [comment]
I think, the 3.77% is misleading!
If you ignore all the trivial scripts, running on single urls or on all urls (or all urls with a certain extension) or on a single site, the real percentage for non-trivial scripts is much higher.
(In my scripts, the percentage is 100%, but that's a horse of another color!)

**Comment by Sxderp** [comment]
You can't just exclude a whole swath of scripts. This was an analysis based on total use, not some subset of use. But even if you did, then you should exclude all the scripts that use regex as glorified globs. I said that about 40% of the scripts that only used regex wildcards. For the purposes of argument, let's say that number is only at 30%.

That would make ~155 scripts with useful / more complex regex. Let's also say that only 40%  (pulling this number out of nowhere) of the 5591 scripts examined are scripts that should be counted using your criteria.

2236 "useful" scripts. That still makes regex ~7%.

**Comment by sizzlemctwizzle** [comment]
@Eselce
The vast majority of userscripts *are* trivial. That is a large part of the appeal. You have a small goal that you want to achieve and you can just throw something together to get the job done. Large and complex userscripts are relatively uncommon. The goal is find a balance between making userscripts simple to create, while providing enough features and flexibility for complex scripts.

In the case of regex, if GM can parse your meta and generate the wrapper then you could just do this in your script. If your script is already "non-trivial", then I don't see how this would significantly complicate things. Certainly it doesn't warrent GM  having to maintain two execution paths. One of which will rarely (if ever) be used by your average GM user.

**Comment by Eselce** [comment]
> The vast majority of userscripts are trivial.

@sizzlemctwizzle I agree. We have a bulk of trivial scripts, a small percentage of non-trivial scripts, most of them with multiple targets or minor wildcards (which can easily be done with @match), but a much higher percentage than 3.77% of these will use regex includes (or excludes, which are less important, maybe simpler?).

And as @Sxderp said, even if the regex includes are only 10% or just a single digit percentage, these are hundreds, thousands, and more scripts already running, do you want to break them without a replacement?

Why not handle complicated checks on `window.location.href` always in the script?
Mostly, because they are shown as triggered (running), even if they return without any impact on the page.
Fine tuning in the `GM` framework is better, but not too good. I hope, there will be a decent interface for it!

**Comment by kekkc** [comment]
Exactly, it should be handled in https://bugzilla.mozilla.org/show_bug.cgi?id=1437098#c4 if this is relevant.

In GM most people don't know about RegEx and therefore don't use them. In addition GM's "RegEx implementation" is fairly limited. Even though I know now about this, it wouldn't help me. No option to use TLD within the RegEx, no option to specify a global search with multiple iterations etc.. If RegEx is considered by Mozilla's API, a trvivial usage like the following should be possible (with an advanced version that parses the complete publicSuffixList).

const TLDs =  'ac(?:ademy)?|accountants|active|actor|ad|ae(?:ro)?|af|ag(?:ency)?|ai(?:rforce)?|al';
const urlRE = new RegExp('^https?:\/\/(?:[^=~?]*[=~?])+(?:' +
'((?:https?|ftp|nntp)(?::|%3a)(?:\/|%2f)(?:\/|%2f)(?:[a-z0-9-]+(?:\\.|%2e)){1,5}(?:' + TLDs + ')(?:(?:\/|%2f|:[0-9]+)(?:[a-z0-9\-._~!$()*+,;=:@\/?]|%[0-9a-f]{2})*)?)|' +
'((?:[a-z0-9-]+(?:\\.|%2e)){2,5}(?:' + TLDs + ')(?:(?:\/|%2f|:[0-9]+)(?:[a-z0-9\-._~!$()*+,;=:@\/?]|%[0-9a-f]{2})*)?)|' +
'((?:[a-z0-9-]+(?:\\.|%2e)){1,5}(?:' + TLDs + ')(?:\/|%2f|:[0-9]+)(?:[a-z0-9\-._~!$()*+,;=:@\/?]|%[0-9a-f]{2})*)|' +
'((aHR0cDov|aHR0cHM6|aHR0cCUz|aHR0cHMl)(?:[a-z0-9+\/=]+))' +
')(?:$|(?:[\-_~!$()*+,;=:@\/&#]|%[0-9a-f]{2}))', 'i');

**Comment by Sxderp** [comment]
@kekkc you missed the code blocks so some of that is garbled. Not really sure what that regex is supposed to look like.

**Comment by kekkc** [comment]
Updated, but the code blocks don't format well, too. However, it was just an example how to get URLs like http://www.yahoo.de/asdf.php?url=www.microsoft.com . I use a more precise version for all links on a page that parses the complete publicSuffixList.

**Comment by trlkly** [comment]
I don't know why people would not know about RegEx compatibility. It's defined in the same place 
[where the rules for globs are defined](https://wiki.greasespot.net/Include_and_exclude_rules), which is a subpage of [where the metadata is defined](https://wiki.greasespot.net/Metadata_Block#.40include). I just think it's largely unnecessary for the types of scripts people tend to make. Most people are fine with the even more limited `@match` rules. 

I'm also not sure I follow the need for global search for this purpose. The point of `@include` and `@exclude` are just to determine whether a script should be run on a particular URL. If it matches once, then it should run (or not run, in the case of `@exclude`). What use case are you envisioning for global searching? 

As for TLD support, I guess it could function as a shorthand, though I suspect that, for most use cases, a shorter RegEx would be sufficient. It's just that globs don't let you choose a shorter set, so using TLDs is often a workable solution. I assume it's mostly used for applying the same script to sites that have versions in multiple countries. 

If it is implemented, I would suggest using `{tld}` as a token that is just replaced with the full TLD RegEx before the script is registered. 

I do note that TLD support is not even listed on the API page. It's only linked in an [orphaned article](https://wiki.greasespot.net/Magic_TLD), without much info. (If it's that short, it should probably just be in the include/exclude rules page.)  I admit, I did not know it existed until it was brought up here.

**Comment by sizzlemctwizzle** [comment]
@kekkc 
>Exactly, it should be handled in https://bugzilla.mozilla.org/show_bug.cgi?id=1437098#c4 if this is relevant.

I agree. It's probably best to try and get as much backwards compatibility supported by the WebExtension API before making any calls on doing workarounds or dropping features.

@trlkly 
> I don't know why people would not know about RegEx compatibility.

It is *relatively* new in the world of userscripting compared to `@include/@exclude` and not supported by `@match`, which has its own pattern.
 
> I admit, I did not know [TLD support] existed until it was brought up here.

I only know about it from seeing the code when I was working on the live metadata feature. It's basically just a wildcard to a list of values (all currently valid TLDs). I agree that regex support likely solves *most* of its use cases. But I've never used the feature myself, so like I said earlier, there could be something that I'm missing (i.e. I honestly don't see point of keeping it around, but it has been in GM for a long time and there seem to be others who really value it).

**Comment by Sxderp** [comment]
> but it has been in GM for a long time and there seem to be others who really value it 

Tld was one thing I forgot to test for in my write-up. I should still have the code somewhere. If not it'd be easy enough to rewrite for tld checks.

**Comment by Sxderp** [comment]
For those not following the Bugzilla ticket. There's at least some serious discussion going into this:
[Evaluate extended matching capabilities for registered userScripts (and/or WebExtensions Content Scripts in general)](https://bugzilla.mozilla.org/show_bug.cgi?id=1445909)

**Comment by Eselce** [comment]
That sounds quite interesting (although it's only a summary). I found the last paragraph quite interesting, as it is in the own interest of Mozilla to support that matching stuff! Interesting thought...

**Comment by Eselce** [comment]
.

**Comment by kekkc** [comment]
So FF 59 is available incl. content scripts now, any plan to make use of it or give users at least the option to use it?

I have many scripts that disable dangerous, security related functions e.g. https://mathiasbynens.github.io/rel-noopener/ that don't work currently, because of the document-start & many other problems of the current implementation.

Anyone planning any forks or any possibility to use Sxderps patch in FF stable? Some ideas would be very welcome.

**Comment by trlkly** [comment]
Getting proper document-start support is definitely something I want
yesterday, but it does look to me like that script is no longer necessary
for Firefox, unless I'm missing something. It lists Firefox as having fixed
the security bug that the script was mitigating.

**Comment by kekkc** [comment]
Tested Tampermonkey. Result:

1. Proper document-start support ( content_scripts key in manifest.json, one has to select "instant" in the settings)
2. Debugging info (browser console shows detailed errors of user scripts)
3. Proper script editor (in GM I cannot even copy my user script to a proper editor, because GM is only copying "-" to the clipboard)
4. Ability to export scripts

Only point where TM is currently behind GM is imporing of LOCAL @require scripts. Currently there's no option to use local scripts with @require in TM. 
However, everything else is usable. In GM, I'm mostly missing  1. & 2., which makes it completely unusable for me. Even worse that these would be easily fixed with programmatic scirpts.

**Comment by Sxderp** [comment]
I'm pretty sure TM lacks on 1. I don't think it can guarantee that a script executes at document start due to message passing. Unless it has implemented the content script API. For 3. the CodeMirror (the editor GM uses) needs to be updated.

**Comment by trlkly** [comment]
My personal experience with TamperMonkey is that, even with its "Instant"
mode workaround, it still always injects after document.head exists. I have
some scripts that injected at document-start but added to the
document.head, and they worked. One script in particular I had tons of
feedback, and telling people to install TamperMonkey in place of
GreaseMonkey 3 was always a successful workaround until I figured out the
problem.

I've also personally tested Instant mode and found plenty of times that
there was a delay. It seems that having an SSD or reinstalling Windows has
made this less of an issue. Granted, this is in Chrome, but I don't see any
mention of doing it differently on Firefox 59 or above in the change
notes.

**Comment by kekkc** [comment]
In contrary to FF, Chrome supports userscript out of the box (no real need for a userscript manager).

However, now even FF ESR is enforcing WebExtensions and still GM is not supporting proper document-start (90% of my scripts don't run & the framework using MutationObservers including the complete userbase from GM3 don't work on most sites).

Is there any update, plan or way forward after 1 year?

**Comment by ExE-Boss** [comment]
@arantius [The `browser.contentScripts` API](https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/contentScripts) is now [supported since Firefox 59](https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/contentScripts#Browser_compatibility).

**Comment by KOLANICH** [comment]
WebNavigation-inserted scripts now are not guaranteed to be executed before page scripts

**Comment by ExE-Boss** [comment]
Also, Firefox provides the [`browser.webRequst.StreamFilter`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/StreamFilter), which allows modifying the HTML payload as it’s received (it’s used in [µBlock Origin](https://github.com/uBlockOrigin)).

**Comment by kekkc** [comment]
Guess @arantius waited on "Implement userScripts.register and execute userScripts js code in isolated sandboxes" https://bugzilla.mozilla.org/show_bug.cgi?id=1437861 ( the browser.contentScripts API was the pre-requisite for this one)

Seems the patch there is revivewed, but needs to be commited now. Hopefully we get this during next week and some test builds here.

**Comment by kekkc** [comment]
So https://bugzilla.mozilla.org/show_bug.cgi?id=1437861 is fixed and implemented in FF64 ;)

**Comment by kekkc** [comment]
FF64 arrived. Any news here or test builds already? Anything still missing from Mozilla?

**Comment by arantius** [comment]
This is a large effort for a small (?) win, so keep your expectations set at an appropriate level.

Or argue for it being a large win?  I don't know what switching gets us that I'd consider anything but small, but maybe I'm missing something.

**Comment by Sxderp** [comment]
While this should be done, I think holding off for iteration two is the better bet. Mostly due to [1509339](https://bugzilla.mozilla.org/show_bug.cgi?id=1509339).

A few things that I can list off.
- Proper injection timing (document-start)
- Initial isolation of API functions and user script. Right now this doesn't gain too much but it sets the needed groundwork for when user scripts can be restricted from WebExtension APIs.
- Less script munging required. APIs can be injected cleanly without all the additional wrapper code.
- [1509339](https://bugzilla.mozilla.org/show_bug.cgi?id=1509339) comes with the potential for more control over matches and includes. Of course this *should not* be used for every script on every page load (@arantius had suggested this in a Bugzilla somewhere). Native WebExtension URL matching has been moved to C code. When possible those mechanisms should be used rather than JS re-implementations.

**Comment by Sxderp** [comment]
> ... that GreaseMonkey injects earlier than it used to, and much earlier than TamperMonkey.
>

The issue is a little more nuanced. Greasemonkey will sometimes catch script injection in time for "document-start" but it's not guaranteed due to asynchronous message passing. The problem you're hitting about being earlier mostly has to do with the script being injected close to document-start. From Mozilla's documentation document-start indicates "DOM is loading." And by that they mean "you probably won't even have `head` or `body` nodes." Apparently GM3 did some munging in order to ensure some minimal DOM was available.

I don't know if the DOM issue can be resolved, but I'd consider adding MutationObservers much less breaking than completely missing your target.

**Comment by kekkc** [comment]
Exactly that's my biggest issue, even with MutationObservers there's currently no reliable method to inject a script before e.g. the title tag (so that the title of a page can be changed). Even worse, it is not executed later, it's simply not executed at all :( Explained in the lenghty original bug since 2013:
https://github.com/greasemonkey/greasemonkey/issues/1849 

And now we've a solid way by Firefox (even if it's ripe only in FF65). The effort was made by @Sxderp already who provided a proof of concept for the initial content script (which is similar and already ready to use). Is this just about the adjustment of this concept and adaption to the current master?

Even if I don't know the code really, I think I could do this. Or would this be waste of time, after dozens of Mozilla & community people provided the grounds here already as a prio 1 mainly for GM, that are currently being discussed if they will be used ?

**Comment by Sxderp** [comment]
I was doing some testing with the `userScripts` API on Nightly. A couple things I've noticed. The UserScript sandboxes do not have the WebExtension API endpoints injected into them (`browser` and `chrome`). Therefore we no longer have to worry about scripts circumventing grants by reimplementing the APIs themselves (right now scripts could cause problems if they found the uuid of another script with the grants). Further, with the `onBeforeScript` event listener we get access to the sandbox object that the script will be executed in which should help with some things.

I'm still testing but looking at the code xrays are still enabled with no way to disable them. And, somewhat problematically the export helpers are not available in the userscript sandbox which does make interacting with the page somewhat difficult. I created a [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1516356) to hopefully address this issue.

**Comment by kekkc** [comment]
Awesome ;) BTW: Luca was fast & had some suggestions in https://bugzilla.mozilla.org/show_bug.cgi?id=1516356

**Comment by Sxderp** [comment]
I have a new proof of concept using the cool APIs available in 66+.
https://github.com/Sxderp/greasemonkey/tree/userscript-api

Unfortunately the backwards compatibility regex (mentioned some post above in this thread) doesn't quite work. There's supposed to be a way to cancel the user script execution in the `onBeforeScript` event but nothing is implemented at this time. As a result, if you have a set of `includes` or `matches` that result in the backward compatibility code being used it'll match all pages.

**Comment by Sxderp** [comment]
Another added benefit is that cross origin requests are no longer leaky ( #2549  ).

**Comment by kekkc** [comment]
This is amazing, thank you so much @Sxderp . Tested and this works flawlessly:

Tested in FF developer 65 & 66 Nightly, xpinstall.signatures.required;false & extensions.webextensions.userScripts.enabled;true, https://github.com/Sxderp/greasemonkey/tree/userscript-api downloaded / zipped / renamed as XPI and loaded in FF, scripts imported via GUI:

1. all scripts are executed reliably again at the start (in contrast to GMs current behavior, where everything is executed AFTER "load", when every malicious script was executed already)
2. Some logging events in browser console, guess those will go away with the final API "<unavailable> user-script-api.js:23:3"
3. Current GMs errors like "security error: content on moz-extension://xxx/ is not allowed to load data from blob:null/xyz." are gone
4. all security issues like https://mathiasbynens.github.io/rel-noopener/ can be successfully prevented via userscripts again (although this seems to be deacted in nightly anyway)
5. errors in userscripts are shown again in the console, hurray (a click shows the position of the error in the code) - party ;)
6. cloneInto() is not yet working, because Mozilla still needs to finalize that
7. Small issue: Global Excludes are currently not respected, maybe related to the  cancel the user script execution issue.

**Comment by ExE-Boss** [comment]
@kekkc
> 4. all security issues like https://mathiasbynens.github.io/rel-noopener/ can successfully be prevented via userscripts again (although this seems to be deacted in nightly anyway)

That’s probably because of [bug&nbsp;1503681](https://bugzilla.mozilla.org/show_bug.cgi?id=1503681).

**Comment by Sxderp** [comment]
> 2. Some logging events in browser console, guess those will go away with the final API " user-script-api.js:23:3"

Yeah, that was me forgetting to remove a debugging line.

> 5. errors in userscripts are shown again in the console, hurray (a click shows the position of the error in the code) - party ;)

I'm thinking about opening a bug to see if this can be further improved. Right now the 'file' is just a string of random alpha numerics. Would be great if we could set fake file names so you know which script is causing the error without having to click on it. Might be a lot of effort for too little gain.

> 6. cloneInto() is not yet working, because Mozilla still needs to finalize that

Interesting, it should work. Granted I didn't test it. I only tested `exportFunction` which worked with Luca's suggestion from the bugzilla ticket.

> 7. Small issue: Global Excludes are currently not respected, maybe related to the cancel the user script execution issue.

Could be, but they should work (untested) if your excludes are globs only and the user script in question is only using includes (as globs) or matches (not both).

**Comment by Sxderp** [comment]
> 7. Small issue: Global Excludes are currently not respected, maybe related to the cancel the user script execution issue.

Oh, there is an issue regarding this. Let us assume that the conditions are right such that the backwards compatibility code is NOT in use.

If you add a new global excludes the script objects in the background will not refresh the cached runtime includes/excludes/matches nor will they reregister themselves. However, if you look at the monkey menu you should see all your scripts move to the "other user scripts" section, as the monkey menu creates new user script objects every time it is opened. The fix is fairly simple, just need to make sure a global refresh is done when global settings are changed (similar to when the plugin is toggled on / off but also refresh the cache). This should only take a few minutes and I'll push a commit.

**Comment by arantius** [comment]
First, huge thanks for all your effort.

> ... script objects in the background will not refresh the cached runtime  ...

Maybe we can make it work, but I'm extremely wary of trying to cache things and keep them in sync as they change, that's a Hard Problem.

**Comment by Sxderp** [comment]
Could just refresh the expressions every time the script is reregistered (which has to be done regardless). Monkey menu does this every time it opens since it recreates the objects. The change would be simple. Remove most of the changes from my latest [commit](https://github.com/Sxderp/greasemonkey/commit/5d17ca4dac609f67430bade60736679445428be2) and make the register command [unconditionally call the refresh](https://github.com/Sxderp/greasemonkey/blob/5d17ca4dac609f67430bade60736679445428be2/src/user-script-obj.js#L469). If this is the case then the refresh [called at the end of an update](https://github.com/Sxderp/greasemonkey/blob/5d17ca4dac609f67430bade60736679445428be2/src/user-script-obj.js#L498) can also be removed.

**Comment by Sxderp** [comment]
I've been thinking about other ways this can benefit. And I'm sure it can lead to simplifying some of the GM APIs. For example I think the XHR API can be rewritten to not send any messages to the background and removed the need for opening the port. I think the same can be done for the open in tab API. Might be able to 'fix' the clipboard API, but I'm not sure about that one. 

Anyway, I really think that switching to this WebExtension API has so much potential for improvement. With the only downside is trying to stay performant by using native matching when possible (since that's done in C code) and maintaining backward compatibility with historic matching mechanisms (which I think should be dropped, but there's probably a large split on that subject).

**Comment by kekkc** [comment]
FF66 is nearing, any plans to include @Sxderp patch within next official GM release?

**Comment by kekkc** [comment]
Finally the userscript API has landed in FF68 and is switched on by default (https://bugzilla.mozilla.org/show_bug.cgi?id=1514809).

Is GM prepared as well?

**Comment by kekkc** [comment]
Update:
- "[META] Implement a userScripts API" is closed an implemented ( https://bugzilla.mozilla.org/show_bug.cgi?id=1437098 )
- exportFunction & cloneInto can be made available to the userscripts to interact with the "real" page ( https://bugzilla.mozilla.org/show_bug.cgi?id=1516356#c1 )
- GM setters & getters were made available

Apart from the advantages mentioned here, this should cover all current features of GM in current FF releases, or are there still showstoppers?

**Comment by kekkc** [comment]
Seems this is a mess. GM and UserScripts seem to be dead. TM & VM don't offer access to DOM (exportfunction / cloneinto) & Mozilla doesn't want to enable it by default: https://github.com/erosman/support/issues/103#issuecomment-565829719

"To protect the user" the user should simply display an alert box with his userscripts (cause there's not much else left without access to the site).

Anyone who has an idea how to use a proper userscript API with FF?

 @Sxderp  
any plans for you to to offer your patch on AMO (or a stripped down version that just enables scripts)?

**Comment by KOLANICH** [comment]
>GM and UserScripts seem to be dead.

Of course it is. Unless someone will start coding instead of complaining in bug trackers.

**Comment by Sxderp** [comment]
@kekkc Good question. I have a few things to say about this but here isn't the place. Send me an email if you want more information. But in short, perhaps, if there's interest.

**Comment by kekkc** [comment]
Would love to get some background, unfortunately there's no longer PM feature in github or mail (and I guess noone wants to include their real mail in the public profile for spammers). You can contact me directly via kekkc@spambog.de

**Comment by kekkc** [comment]
I'm so sick of it. This POC is based from the initial one from one year ago & extended by Luca's suggestion for exportFunction / cloneInto support. This saves currently only one script, but runs at document-start, correct error messages and all the fancy stuff that we want here.

If anyone wants to support, let's include support for multiple scripts and bring it to AMO for easy installation and forget this crap here ongoing since 1 year (all the while with a working POC in the background, all credits to https://bugzilla.mozilla.org/user_profile?user_id=609792 (unfortunately this great POC didn't even make it to the official webextension examples https://github.com/mdn/webextensions-examples 

*sry for the rant, I'm quiet now, once and for all*

https://bugzilla.mozilla.org/attachment.cgi?id=9033244
https://bugzilla.mozilla.org/show_bug.cgi?id=1516356#c1

[user-script-register-example-v2.zip](https://github.com/greasemonkey/greasemonkey/files/4131469/user-script-register-example-v2.zip)

**Comment by arantius** [comment]
> sry for the rant, I'm quiet now, once and for all

It's really hard to respond to negativity positively.  It would be more constructive to calm down and reword, than to tack on an apology.

> runs at document-start, correct error messages and all the fancy stuff that we want here.

Besides not knowing what "all the fancy stuff" is, that sounds pretty good.
